File: Behavioral - Mediator/Typescript/mediator/concrete_mediator.ts
import { Mediator } from '../interface/interface.ts'
import { Component1 } from '../components/component1.ts'
import { Component2 } from '../components/component2.ts'


export class ConcreteMediator implements Mediator {
  private component1: Component1;

  private component2: Component2;

  constructor(c1: Component1, c2: Component2) {
      this.component1 = c1;
      this.component1.setMediator(this);
      this.component2 = c2;
      this.component2.setMediator(this);
  }

  public notify(sender: object, event: string): void {
      if (event === 'A') {
          console.log('Mediator reacts on A and triggers following operations:');
          this.component2.doC();
      }

      if (event === 'D') {
          console.log('Mediator reacts on D and triggers following operations:');
          this.component1.doB();
          this.component2.doC();
      }
  }
}

File: Behavioral - Mediator/Python/mediator/concrete_mediator.py
from component.component1 import Component1
from component.component2 import Component2
from interface.interface import Mediator


class ConcreteMediator(Mediator):
  def __init__(self, component1: Component1, component2: Component2) -> None:
    self._component1 = component1
    self._component1.mediator = self
    self._component2 = component2
    self._component2.mediator = self

  def notify(self, sender: object, event: str) -> None:
    if event == "A":
      print("Mediator reacts on A and triggers following operations:")
      self._component2.do_c()
    elif event == "D":
      print("Mediator reacts on D and triggers following operations:")
      self._component1.do_b()
      self._component2.do_c()